File: program11.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 100
int heap[MAX], size = 0;

// swap
void swap(int *a, int *b) {
    int t = *a;
    *a = *b;
    *b = t;
}

// Min-Heap insert
void insertMin(int val) {
    heap[++size] = val;
    int i = size;
    while (i > 1 && heap[i / 2] > heap[i]) {
        swap(&heap[i], &heap[i / 2]);
        i /= 2;
    }
}

// Max-Heap insert
void insertMax(int val) {
    heap[++size] = val;
    int i = size;
    while (i > 1 && heap[i / 2] < heap[i]) {
        swap(&heap[i], &heap[i / 2]);
        i /= 2;
    }
}

// display
void display() {
    if (size == 0) {
        printf("Heap empty\n");
        return;
    }
    for (int i = 1; i <= size; i++)
        printf("%d ", heap[i]);
    printf("\n");
}

int main() {
    int ch, val, type;
    while (1) {
        printf("\n1.Insert\n2.Display\n3.Exit\nEnter choice: ");
        scanf("%d", &ch);
        switch (ch) {
            case 1:
                printf("Enter 1 for Min-Heap, 2 for Max-Heap: ");
                scanf("%d", &type);
                printf("Enter value: ");
                scanf("%d", &val);
                size = 0; // reset for new heap type
                if (type == 1)
                    insertMin(val);
                else
                    insertMax(val);
                break;
            case 2:
                printf("Heap elements: ");
                display();
                break;
            case 3:
                exit(0);
            default:
                printf("Invalid choice\n");
        }
    }
}

File: program11.py
import heapq

class MinHeap:
    def __init__(self):
        self.heap = []

    def insert(self, value):
        heapq.heappush(self.heap, value)

    def delete(self):
        if not self.heap:
            return None
        return heapq.heappop(self.heap)

    def display(self):
        print("Min Heap:", self.heap)


class MaxHeap:
    def __init__(self):
        self.heap = []

    def insert(self, value):
        heapq.heappush(self.heap, -value)  # push negative to simulate max heap

    def delete(self):
        if not self.heap:
            return None
        return -heapq.heappop(self.heap)

    def display(self):
        # convert back to positive for displaying
        print("Max Heap:", [-x for x in self.heap])


# -------- Main Program --------
if __name__ == "__main__":
    min_heap = MinHeap()
    max_heap = MaxHeap()

    while True:
        print("\n--- HEAP OPERATIONS ---")
        print("1. Insert in Min Heap")
        print("2. Delete from Min Heap")
        print("3. Display Min Heap")
        print("4. Insert in Max Heap")
        print("5. Delete from Max Heap")
        print("6. Display Max Heap")
        print("7. Exit")

        choice = int(input("Enter your choice: "))

        if choice == 1:
            value = int(input("Enter value: "))
            min_heap.insert(value)

        elif choice == 2:
            deleted = min_heap.delete()
            print("Deleted:", deleted)

        elif choice == 3:
            min_heap.display()

        elif choice == 4:
            value = int(input("Enter value: "))
            max_heap.insert(value)

        elif choice == 5:
            deleted = max_heap.delete()
            print("Deleted:", deleted)

        elif choice == 6:
            max_heap.display()

        elif choice == 7:
            print("Exiting...")
            break

        else:
            print("Invalid choice! Try again.")
